src/lsptoolshost/projectRestore/restore.ts

```diff
@@ -18,31 +18,63 @@ import { getCSharpDevKit } from '../../utils/getCSharpDevKit';
 
 let _restoreInProgress = false;
 
-export function registerRestoreCommands(context: vscode.ExtensionContext, languageServer: RoslynLanguageServer) {
-    if (getCSharpDevKit()) {
-        // We do not need to register restore commands if using C# devkit.
-        return;
-    }
-    const restoreChannel = vscode.window.createOutputChannel(vscode.l10n.t('.NET NuGet Restore'));
+export function registerRestoreCommands(
+    context: vscode.ExtensionContext,
+    languageServer: RoslynLanguageServer,
+    csharpOutputChannel: vscode.LogOutputChannel
+) {
     context.subscriptions.push(
         vscode.commands.registerCommand('dotnet.restore.project', async (_request): Promise<void> => {
-            return chooseProjectAndRestore(languageServer, restoreChannel);
+            if (getCSharpDevKit()) {
+                csharpOutputChannel.debug(
+                    "[.NET Restore] Not handling command 'dotnet.restore.project' from C# extension, because C# Dev Kit is expected to handle it."
```

Comment:
What if we just didn't register these commands if using CDK (also remove the enablement from the package.json) ?

Could still hook the restore request.

Reply:
I think I did this in a previous revision, but, an integration test failed IIRC due to registering the command, when the enablement expression in package.json was false. So it's necessary to have the enablement as well as the `registerCommand` in sync, I guess.

What I want is just to be certain that none of these commands will ever silently do nothing.

One thing which is confusing here is; does this function re-run when CDK is loaded/unloaded, then "Reload Extensions" is run? That does not seem to be the case, and neither does the "Restart Language Server" command I was prompted to run when loading/unloading CDK.

I'll have to jump in the debugger and check.

Reply:
OK. `registerRestoreCommands` is called again when CDK changes from loaded->unloaded, but not when it changes from unloaded->loaded. So maybe enablements settings cause the commands to get filtered out for the unloaded->loaded case?

It seems like it is fine to adjust things to condition the `registerCommand` calls based on whether CDK is loaded. Just the `languageServer.registerOnRequest` call needs to happen regardless of CDK state. And in that case it seems good to delay the CDK check as the PR is currently doing.

Reply:
> OK. `registerRestoreCommands` is called again when CDK changes from loaded->unloaded

The fact that I had to reload the window to repro the restore bug, after unloading CDK, makes it seem like the `languageServer.registerOnRequest` callback is not unregistered when CDK is unloaded. Seems funky. Does this mean that if we kept reloading->unloading CDK, we would register more and more callbacks?

Reply:
Repeatedly calling `registerOnRequest` doesn't seem to cause problems. For example, if we send the message the handler is registered to handle, we don't get duplicate log messages indicating duplicate runs of the handler. So I don't think there is anything further that needs to be fixed.